### PR TITLE
Separate typed into modules

### PR DESCRIPTION
**Pull request by chbatey**
Creates the following modules and renames packages to match:

`akka.actor.typed`
`akka.persistence.typed`
`akka.cluster.typed`
`akka.cluster.sharding.typed`

Differences from untyped:
* DData and cluster singleton are inside cluster. DData is in the package `akka.cluster.dddata.typed`

Refs #23632

### CODE REVIEW

File: akka-actor-typed-tests/src/test/resources/reference.conf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+akka.typed {
+  # for the akka.actor.ExtensionSpec
+  library-extensions += "akka.actor.typed.InstanceCountingExtension"
+}
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/resources/reference.conf]: add newline, can be of importance in conf files

File: akka-cluster-sharding-typed/src/main/resources/reference.conf [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# TODO: move these out somewhere else when doing #23632
```
Comment (patriknw) [inline akka-cluster-sharding-typed/src/main/resources/reference.conf]: done?

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by chbatey** [comment]
Most of this is mechanical, I'd suggest having a look through the build.sbt to make sure i've got the module names right + the dependencies between the modules

File: build.sbt [hunk 3]
```diff
@@ -386,21 +380,59 @@ lazy val typed = akkaModule("akka-typed")
...
-    typedTestkit % "compile->compile;test->provided;test->test",
-    // the provided dependencies
-    persistence % "compile->compile;test->test",
-    cluster % "test->test",
+    actorTyped,
+    cluster,
     clusterTools,
+    distributedData,
+    persistence % "provided->test",
+    persistenceTyped % "provided->test",
```
Comment (patriknw) [inline build.sbt]: why dependencies to persistence from cluster?

File: build.sbt [hunk 4]
```diff
@@ -386,21 +380,59 @@ lazy val typed = akkaModule("akka-typed")
...
+  .settings(Protobuf.importPath := Some(baseDirectory.value / ".." / "akka-remote" / "src" / "main" / "protobuf" ))
+  .disablePlugins(MimaPlugin)
+
+
+lazy val typedTestkit = akkaModule("akka-testkit-typed")
+  .dependsOn(actorTyped, testkit % "compile->compile;test->test")
+  .settings(AutomaticModuleName.settings("akka.testkit.typed"))
+  .disablePlugins(MimaPlugin)
+
+lazy val actorTypedTests = akkaModule("akka-actor-typed-tests")
```
Comment (patriknw) [inline build.sbt]: something missing for saying that this should not be published to maven central?

**Review by patriknw — COMMENTED** [review]
looked closer at build.sbt and might have found something there

File: build.sbt (hunk 3 above)
Comment (chbatey) [reply build.sbt]: It is just for test, as we test having a persistent cluster singleton

File: build.sbt (hunk 4 above)
Comment (chbatey) [reply build.sbt]: That is a couple of lines below, `enablePlugins(NoPublish)`

File: akka-cluster-sharding-typed/src/main/resources/reference.conf (hunk 2 above)
Comment (chbatey) [reply akka-cluster-sharding-typed/src/main/resources/reference.conf]: :+1:

**Review by johanandren — APPROVED** [review]
LGTM

**Pull request merged**